pkg/antctl/command_list.go

```diff
@@ -0,0 +1,170 @@
+		resources[groupPath] = append(resources[groupPath], metav1.APIResource{
+			Name:         def.Use,
+			SingularName: def.Use,
+			Kind:         def.Use,
+			Namespaced:   false,
+			Group:        def.GroupVersion.Group,
+			Version:      def.GroupVersion.Version,
+			Verbs:        metav1.Verbs{"get"},
+		})
+	}
+	// Setup up discovery handlers for command handlers.
+	for groupPath, resource := range resources {
+		mux.HandleFunc(groupPath, func(w http.ResponseWriter, r *http.Request) {
+			w.Header().Set("Content-Type", "application/json")
+			l := metav1.APIResourceList{
+				TypeMeta:     metav1.TypeMeta{Kind: "APIResourceList", APIVersion: metav1.SchemeGroupVersion.Version},
+				GroupVersion: systemGroup.Version,
+				APIResources: resource,
+			}
+			jsonResp, err := json.MarshalIndent(l, "", "  ")
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+			_, err = w.Write(jsonResp)
+			if err != nil {
+				klog.Errorf("Error when responding APIResourceList: %v", err)
+				w.WriteHeader(http.StatusInternalServerError)
+			}
+		})
+	}
+}
+
+func (cl *commandList) applyFlagsToRootCommand(root *cobra.Command) {
+	defaultKubeconfig := filepath.Join(homedir.HomeDir(), ".kube", "config")
+	root.PersistentFlags().BoolP("verbose", "v", false, "enable verbose output")
+	root.PersistentFlags().StringP("kubeconfig", "k", defaultKubeconfig, "absolute path to the kubeconfig file")
+	root.PersistentFlags().DurationP("timeout", "t", 0, "time limit of the execution of the command")
+}
+
+// ApplyToRootCommand applies the commandList to the root cobra applySubCommandToRoot, it applies
+// each commandDefinition of it to the root applySubCommandToRoot as a sub-applySubCommandToRoot.
+func (cl *commandList) ApplyToRootCommand(root *cobra.Command, isAgent bool, inPod bool) {
+	client := &client{
+		inPod: inPod,
+		codec: cl.codec,
+	}
+	for _, groupCommand := range groupCommands {
+		root.AddCommand(groupCommand)
+	}
+	for i := range cl.definitions {
+		def := cl.definitions[i]
+		if (def.Agent != isAgent) && (def.Controller != !isAgent) {
+			continue
+		}
+		def.applySubCommandToRoot(root, client, isAgent)
+	}
+	cl.applyFlagsToRootCommand(root)
+	root.PersistentPreRunE = func(cmd *cobra.Command, args []string) error {
+		enableVerbose, err := root.PersistentFlags().GetBool("verbose")
+		if err != nil {
+			return err
+		}
+		err = flag.Set("logtostderr", fmt.Sprint(enableVerbose))
+		if err != nil {
+			return err
+		}
+		err = flag.Set("alsologtostderr", fmt.Sprint(enableVerbose))
+		if err != nil {
+			return err
+		}
+		if enableVerbose {
+			err := flag.Set("v", fmt.Sprint(math.MaxInt32))
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+	renderDescription(root, isAgent)
+}
+
+// validate checks the validation of the commandList.
+func (cl *commandList) validate() []error {
+	var errs []error
+	if len(cl.definitions) == 0 { // must has at least one applySubCommandToRoot
```

Comment:
```suggestion
	if len(cl.definitions) == 0 { // must have at least one applySubCommandToRoot
```
But I think the code is self-explaining already, so do L151, no need to have those comments.